### PR TITLE
Adding Nth-polynomials

### PR DESCRIPTION
**Pull request by thomas-lamiaux**
Adding nth-polynomials using a new definition

### CODE REVIEW

**Comment by felixwellen** [comment]
Interesting! Did you see Cubical/Algebra/CommAlgebra/Instances/FreeCommAlgebra.agda?
I think that is related to what you do. It seems you are a bit more economical with the constructors of your type of polynomials.

Btw: You can mark your pull-request as 'draft'. This is a good way to indicate, that it is work in progress.

**Comment by thomas-lamiaux** [comment]
> Interesting! Did you see Cubical/Algebra/CommAlgebra/Instances/FreeCommAlgebra.agda? I think that is related to what you do. It seems you are a bit more economical with the constructors of your type of polynomials.
 
I didn't see it even though I am currently working with Anders, I think that they originated very differently.
Basically we were looking for a good definition of the direct-sum. Working with one we've come up with, I have realized that I could define nth-polynomials easily and it would simplify some latter proof. 

Plus, I am not a big fan of the current one based on the :: operation. I prefer this one base on "vectorial space". With the difference that the action on the basis is internalized - which comes from the direct sum that is dependent - and seems to facilitate definition and proofs. 
Indeed basically it suffices to specify the behavior for aX^n, to define functions such as the product so all structures rises and that simplify the structure.
 
> Btw: You can mark your pull-request as 'draft'. This is a good way to indicate, that it is work in progress.

Indeed I forgot to do it, it's done thanks

**Comment by thomas-lamiaux** [comment]
Would adding a syntactic sugar like X^n1...X^nk for base [n1, ... nk] 1 be interesting ? 

(I don't really know how to do so for any k)

File: Cubical/Algebra/Polynomials/Nth-polynomials/Base.agda [hunk 1]
```diff
@@ -0,0 +1,137 @@
+{-# OPTIONS --safe #-}
+module Cubical.Algebra.Polynomials.Nth-polynomials.Base where
+
+open import Cubical.Foundations.Everything
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Base.agda:4]: Don't import Everything

File: Cubical/Algebra/Polynomials/Nth-polynomials/Base.agda [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+open import Cubical.Data.Nat renaming (_+_ to _+n_)
+open import Cubical.Data.Vec
+
+open import Cubical.Algebra.Group
+open import Cubical.Algebra.AbGroup
+open import Cubical.Algebra.Ring
+open import Cubical.Algebra.CommRing
+
+private variable
+  l l' : Level
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Base.agda:16]: \ell

File: Cubical/Algebra/Polynomials/Nth-polynomials/CommRing-Structure.agda [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+  _Poly*_ : Poly A' n → Poly A' n → Poly A' n
+  _Poly*_ = -- Induction Left Argument
+            Poly-Rec-Set.f A' n (Poly A' n → Poly A' n)
+            (λ f g p q i j Q → trunc (f Q) (g Q) (λ X → p X Q) (λ X → q X Q) i j )
+            (λ Q    → 0P)
+            (λ v a  → -- Induction Right Argument
+                       Poly-Rec-Set.f A' n (Poly A' n) trunc
+                       0P
+                       (λ v' a' → base (v +n-vec v') (a · a'))
+                       (λ PS QS → PS Poly+ QS)
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/CommRing-Structure.agda:119]: Poly+

File: Cubical/Algebra/Polynomials/Nth-polynomials/CommRing-Structure.agda [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+  open Nth-Poly-structure A' n
+
+  PolyCommRing : CommRing l
+  fst PolyCommRing = Poly A' n
+  0r (snd PolyCommRing) = 0P
+  1r (snd PolyCommRing) = 1P
+  _+_ (snd PolyCommRing) = _Poly+_
+  _·_ (snd PolyCommRing) = _Poly*_
+  - snd PolyCommRing = Poly-inv
+  is-set (isSemigroup (isMonoid (isGroup (IsRing.+IsAbGroup (IsCommRing.isRing (isCommRing (snd PolyCommRing))))))) = trunc
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/CommRing-Structure.agda:239]: Use makeIsCommRing

File: Cubical/Algebra/CommRing/Instances/Polynomials.agda [hunk 5]
```diff
@@ -4,7 +4,7 @@ module Cubical.Algebra.CommRing.Instances.Polynomials where
 
 open import Cubical.Foundations.Prelude
 open import Cubical.Algebra.CommRing
-open import Cubical.Algebra.Polynomials
+open import Cubical.Algebra.Polynomials.Polynomials
```
Comment (mortberg) [inline Cubical/Algebra/CommRing/Instances/Polynomials.agda:7]: Rename to `Univariate`

File: Cubical/Algebra/Polynomials/Nth-polynomials/Base.agda [hunk 6]
```diff
@@ -0,0 +1,137 @@
+{-# OPTIONS --safe #-}
+module Cubical.Algebra.Polynomials.Nth-polynomials.Base where
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Base.agda:2]: Rename to `Multivariate`

File: Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Poly1-Poly.agda [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+  open Nth-Poly-structure A' 1
+
+-- Notation P, Q, R... for Poly 1
+-- x, y, w... for Poly:
+-- a,b,c... for A
+
+
+
+-----------------------------------------------------------------------------
+-- lemma translation Xn
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Poly1-Poly.agda:60]: Move to properties file for univariate polynomials

Comment (mortberg) [reply Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Poly1-Poly.agda:60]: Also split into base+properties

File: Cubical/Algebra/Polynomials/Nth-polynomials/CommRing-Structure.agda [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+               (λ v a → Poly-Ind-Prop.f A' n _ (λ _ → trunc _ _)
+                          refl
+                          (λ v' a' → cong₂ base (+n-vec-comm v v') (·Comm a a'))
+                          (λ {U V} ind-U ind-V → cong₂ _Poly+_ ind-U ind-V))
+               λ {U V} ind-U ind-V Q → ((cong₂ _Poly+_ (ind-U Q) (ind-V Q)) ∙ sym (Poly*-Rdist Q U V))
+
+
+-----------------------------------------------------------------------------
+
+module _ (A' : CommRing l) (n : ℕ) where
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/CommRing-Structure.agda:216]: Move to instantiation to Cubical.Algebra.CommRing.Instances

File: Cubical/Algebra/Polynomials/Nth-polynomials/CommRing-Structure.agda [hunk 9]
```diff
@@ -0,0 +1,250 @@
+{-# OPTIONS --safe --experimental-lossy-unification #-}
+module Cubical.Algebra.Polynomials.Nth-polynomials.CommRing-Structure where
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/CommRing-Structure.agda:2]: Rename to Properties

File: Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Poly0-A.agda [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+open import Cubical.Algebra.AbGroup
+open import Cubical.Algebra.Ring
+open import Cubical.Algebra.CommRing
+open import Cubical.Algebra.Polynomials.Polynomials
+
+open import Cubical.Algebra.Polynomials.Nth-polynomials.Base
+open import Cubical.Algebra.Polynomials.Nth-polynomials.CommRing-Structure
+
+private variable
+  l l' : Level
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Poly0-A.agda:22]: delete l'

File: Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Comp-Poly.agda [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+  sep-vec-snd : (k l : ℕ) → (v : Vec ℕ k) → (v' : Vec ℕ l) → snd (sep-vec k l (v ++ v')) ≡ v'
+  sep-vec-snd zero l [] v' = refl
+  sep-vec-snd (suc k) l (x ∷ v) v' = sep-vec-snd k l v v'
+
+  sep-vec-id : (k l : ℕ) → (v : Vec ℕ (k +n l)) → fst (sep-vec k l v) ++ snd (sep-vec k l v) ≡ v
+  sep-vec-id zero l v = refl
+  sep-vec-id (suc k) l (x ∷ v) = cong (λ X → x ∷ X) (sep-vec-id k l v)
+
+  rep-concat : (k l : ℕ) → {B : Type l'} → (b : B) → replicate {_} {k} {B} b ++ replicate {_} {l} {B} b ≡ replicate {_} {k +n l} {B} b
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Comp-Poly.agda:53]: long line

File: Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Comp-Poly.agda [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+  rep-concat zero l b = refl
+  rep-concat (suc k) l b = cong (λ X → b ∷ X) (rep-concat k l b)
+
+  +n-vec-concat : (k l : ℕ) → (v w : Vec ℕ k) → (v' w' : Vec ℕ l) → (v Mr.+n-vec w) ++ (v' Mr.+n-vec w') ≡ (v ++ v') Mr.+n-vec (w ++ w')
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Comp-Poly.agda:57]: long

File: Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Comp-Poly.agda [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+  private
+    A = fst A'
+  open CommRingStr (snd A')
+
+  module Mr  = Nth-Poly-structure A' m
+  module N+Mr = Nth-Poly-structure A' (n +n m)
+  module N∘Mr  = Nth-Poly-structure (PolyCommRing A' m) n
+
+-----------------------------------------------------------------------------
+-- Some lemma for the traduction
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Comp-Poly.agda:35]: Move to Vec, maybe generalize?

File: Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Induced-Poly.agda [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+private variable
+  l l' : Level
+
+
+-----------------------------------------------------------------------------
+-- Lift
+
+open IsRingHom
+
+lift-poly : (A' B' : CommRing l) → (f : CommRingHom A' B') → (n : ℕ) → CommRingHom (PolyCommRing A' n) (PolyCommRing B' n)
```
Comment (mortberg) [inline Cubical/Algebra/Polynomials/Nth-polynomials/Equiv/Induced-Poly.agda:32]: `makeCommRingHomPoly`

also A' and B' can be in different universes?

File: Cubical/Algebra/Ring/Properties.agda [hunk 15]
```diff
@@ -221,6 +222,28 @@ module RingEquivs where
   fst (compRingEquiv f g) = compEquiv (f .fst) (g .fst)
   snd (compRingEquiv f g) = compIsRingEquiv {g = g .fst} {f = f .fst} (g .snd) (f .snd)
 
+  isRingHomInv : {A' B' : Ring ℓ} → (e : RingEquiv A' B') → IsRingHom (snd B') (invEq (fst e)) (snd A')
```
Comment (mortberg) [inline Cubical/Algebra/Ring/Properties.agda:225]: Different universes

File: Cubical/Algebra/Ring/Properties.agda [hunk 16]
```diff
@@ -221,6 +222,28 @@ module RingEquivs where
...
+                                   g (snd B' ._·_ (f (g x)) (f (g y))) ≡⟨ cong g (sym (pres· fcrh (g x) (g y))) ⟩
+                                   g (f (snd A' ._·_ (g x) (g y)))     ≡⟨ retEq et (snd A' ._·_ (g x) (g y)) ⟩
+                                   snd A' ._·_ (g x) (g y)  ∎)
+               where
+               et = fst e
+               f = fst et
+               fcrh = snd e
+               g = invEq et
+
+  invEquivRing : {A B : Ring ℓ} → RingEquiv A B → RingEquiv B A
```
Comment (mortberg) [inline Cubical/Algebra/Ring/Properties.agda:242]: universes

File: Cubical/Algebra/CommRing/Properties.agda [hunk 17]
```diff
@@ -177,6 +177,10 @@ module CommRingEquivs where
                                                            {B = CommRing→Ring B}
                                                            {C = CommRing→Ring C}
 
+ invCommRingEquiv : (A B : CommRing ℓ) → CommRingEquiv A B → CommRingEquiv B A
```
Comment (mortberg) [inline Cubical/Algebra/CommRing/Properties.agda:180]: different universes

**Review by mortberg — CHANGES_REQUESTED** [review]
Also please remove trailing whitespace

**Comment by mortberg** [comment]
There's still some trailing whitespace:

```
~/.cabal/bin/fix-whitespace --check
[ Violation detected ] Cubical/Algebra/Polynomials/Univariate/Properties.agda
```

**Review by mortberg — APPROVED** [review]

**Pull request merged**